File: Algorithms/CPP/20 Valid Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        std::stack<char> parentheses;
        for (auto c : s) {
            if (c == ')' || c == ']' || c == '}') {
                if (parentheses.empty()) {
                    return false;
                }
                if ((c == ')' && parentheses.top() == '(') 
                || (c == ']' && parentheses.top() == '[')
                || (c == '}' && parentheses.top() == '{')) {
                    parentheses.pop();
                }
                else {
                    return false;
                }
            }
            else {
                parentheses.push(c);
            }
        }

        return parentheses.empty();
    }
};

File: Algorithms/Python/20 Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        parentheses = []
        for c in s:
            if c == ')' or c == ']' or c == '}':
                if len(parentheses) == 0:
                    return False
                elif (c == ')' and parentheses[-1] == '(') or (c == ']' and parentheses[-1] == '[') or (c == '}' and parentheses[-1] == '{'):
                    parentheses.pop()
                else:
                    return False
            else:
                parentheses.append(c)

        return len(parentheses) == 0
